stlab/future.hpp

```diff
@@ -900,327 +933,261 @@ auto when_all(S s, F f, future<Ts>... args) {
-            , _holds(_remaining)
-        {}
 
-        std::atomic_size_t                    _remaining{0};
+    template <typename CR, typename F>
+    struct when_range_context_base : CR
+    {
+        std::atomic_size_t                    _remaining;
+        std::atomic_bool                      _error_happened{ false };
         std::vector<future<void>>             _holds;
-        std::mutex                            _mutex;
         boost::optional<std::exception_ptr>   _error;
-        size_t                                _index{0};
         packaged_task<>                       _f;
 
-        void failure(std::exception_ptr error) {
-            // we already have a result
-            if (_remaining == 0) {
-                return;
-            }
-            // only the last error is of any interest
-            if (_remaining == 1) {
-                _error = std::move(error);
-                _f();
+        when_range_context_base(F f, size_t s)
+            : CR(std::move(f), s)
+            , _remaining(s)
+            , _holds(_remaining)
+        {}
+
+        auto execute() {
+            if (this->_error) {
+                std::rethrow_exception(this->_error.get());
             }
-            --_remaining;
+            return CR::operator()();
         }
     };
 
-    template <typename F, typename I, typename Input>
-    struct when_any_range_context : when_any_range_context_base<F, I> {
-        when_any_range_context(I first, I last)
-            : when_any_range_context_base<F, I>(first, last)
+
+    template <typename CR, typename F>
+    struct when_all_range_context : when_range_context_base<CR, F>
+    {
+        when_all_range_context(F f, size_t s)
+            : when_range_context_base<CR, F>(std::move(f), s)
         {}
-        
-        void done(Input r, size_t index) {
-            if (when_any_range_context_base<F,I>::_remaining == 0)
+
+        template<typename FF>
+        void done(FF&& f, size_t index) {
+            this->apply(std::forward<FF>(f), index);
+            if (--this->_remaining == 0) this->_f();
+        }
+
+        void failure(std::exception_ptr error, size_t) {
+            bool current_error_happened = this->_error_happened.load();
+            if (current_error_happened)
                 return;
-            {
-                std::unique_lock<std::mutex> lock(when_any_range_context_base<F, I>::_mutex);
-                if (when_any_range_context_base<F, I>::_remaining != 0) {
-                    when_any_range_context_base<F, I>::_remaining = 0;
-                    _result = std::move(r);
-                    when_any_range_context_base<F, I>::_index = index;
-                    for (auto& h : this->_holds) {
-                        h.cancel_try();
-                    }
-                }
-                else {
-                    return;
-                }
+
+            if (this->_error_happened.compare_exchange_strong(current_error_happened, true)) {
```

Comment:
Same comment as above - just expect false here, the additional load doesn't help.